### ISSUE TITLE
Apple Deprecating OpenCL/OpenGL on iOS & macOS

### ISSUE THREAD

**Issue by mbrockman1** [kivy/kivy#5789]
Obviously not an immediate issue due to deprecation and not immediate removal, but what does this mean for Kivy/Apple?

**Comment by tshirtman** [comment]
It might become necessary to investigate in the future, if there are, like ANGLE on windows, free implementation of opengl/ES on top of Metal. I know there are commercial ones, but i'm not aware of free ones.

Rewriting kivy for Metal directly, or writting such a compatibility layer ourselves would certainly be a very big effort, and i don't see it happening without a significant funding.

**Comment by stefanhuber** [comment]
Using MoltenVK might be an option: https://github.com/KhronosGroup/MoltenVK

**Comment by mischievous** [comment]
Dumb question.... If I wanted to swap out the opengl for metal or just a c++ osx window (no cocoa) would it be straight forward? I see backends for "debug", "opengl", "mock", and "sdl2".  Would it be as simple as adding a backend for metal and then generating the libraries on osx?

**Comment by tshirtman** [comment]
it's not trivial, but there has been work done at least for kivy-ios to be able to use a fork of Angle that supports the Metal api, and it works, https://github.com/kivy/kivy-ios/pull/448 though it requires more testing, i think maybe such changes should be integrated in kivy, not just kivy-ios, to support OSX, but i didn't look in details into what that entailed.

**Comment by rpopovici** [comment]
@tshirtman what is the staus of this issue? From the looks of it, the latest official angle version(https://github.com/google/angle) has complete support for OpenGL ES2..

**Comment by tshirtman** [comment]
That's great to hear, i didn't follow on this issue and i don't own a mac, but as the test with metalAngle apparently had good results, surely it should be possible to now use the official angle with kivy on osx. It just needs someone with osx and some time to do it, contributions are welcome :).

**Comment by rpopovici** [comment]
@tshirtman I would do it, but I am new to kivy and I don't know exactly what needs to be done. It appears that kivy_deps.angle needs to be built for mac os and the binaries bundled with the kivy whl. Other than that this line need to be removed https://github.com/kivy/kivy/blob/c8698d40d2d2d9d0b9df8d65b5b393551b870b80/kivy/graphics/cgl.pyx#L108 and it should work..

**Comment by misl6** [comment]
At this time, `MetalANGLE` has only been tested over iOS, and it needed some changes over `SDL2`, but seems pretty easy to port it on MacOS.
See https://github.com/kivy/kivy-ios/pull/448 for reference.

Referring to ANGLE vs MetalANGLE, here you can find a detailed report:
https://github.com/kakashidinho/metalangle#differences-between-metalangle-and-googles-angle

**Comment by rpopovici** [comment]
@misl6 are you sure the information from https://github.com/kakashidinho/metalangle#differences-between-metalangle-and-googles-angle is up to date? According to the official repo https://github.com/google/angle OpenGL ES2 is "complete" for Metal render.

Anyway, I managed to use angle with MacOS. The easiest way to do it is to extract GLES related libraries "Chromium.app/Contents/Frameworks/Chromium Framework.framework/Libraries" from chrome browser or chromium app. Copy these libs into /usr/local/lib. Then remove the warning from https://github.com/kivy/kivy/blob/c8698d40d2d2d9d0b9df8d65b5b393551b870b80/kivy/graphics/cgl.pyx#L108 and reinstall kivy from source.
Don't forget to set the backend in your app
```
os.environ["KIVY_GL_BACKEND"] = "angle_sdl2"
```
Run your monkey head example file and you should see this:
![Screenshot 2021-02-02 at 15 14 14](https://user-images.githubusercontent.com/2832454/106607782-5d3bb700-656c-11eb-8750-073ec8fe71b8.png) and the head spinning, of course. 

Now, I have no idea if under the hood angle is using OpenGL or Metal and I don’t know how to check it. Probably if I build angle from sources and make sure that only metal is selected as a backend..

**Comment by matham** [comment]
Yeah, I expected that on osx (unlike ios), simply removing `if platform != "win32":` and making sure angle is installed as @rpopovici indicated should be enough. But perhaps this leads to issues down the line with sdl2?

**Comment by misl6** [comment]
Is completed, but not the Platform support + iOS looks planned but not implemented.

On iOS, SDL2 was linked against OpenGLES, We should look at their code and check if it's the same on MacOS

**Comment by rpopovici** [comment]
I managed to compile angle with metal instead of opengl
The trick was to turn off opengl and activate metal in the args.gn config file:

```
is_clang = true
is_debug = false
angle_enable_gl = false
angle_enable_metal = true
```

Don't turn off vulkan backend or directx. It won't compile!

![Screenshot 2021-02-03 at 00 05 05](https://user-images.githubusercontent.com/2832454/106669793-d9f28380-65b4-11eb-8a68-a148b9788af2.png)

**Comment by rpopovici** [comment]
Worth mentioning that there is Zink now with MacOS support for MoltenVK
https://www.phoronix.com/scan.php?page=news_item&px=Zink-macOS-OpenGL-Vulkan
https://github.com/mesa3d/mesa/tree/master/src/gallium/drivers/zink

**Comment by rpopovici** [comment]
Starting with today's chromium build, ANGLE_DEFAULT_PLATFORM env var can be used to set the default render for angle. Building angle from sources is not required anymore, you can extract the GLES libs from chromium bundle.

```
import os
os.environ['KIVY_WINDOW'] = "sdl2"
os.environ['ANGLE_DEFAULT_PLATFORM'] = "metal"
os.environ["KIVY_GL_BACKEND"] = "angle_sdl2"
```

**Comment by misl6** [comment]
Fixed via #8534 🥳 🎉 

See full story here: #8572
